### PR TITLE
Feature/pacemaker interface

### PR DESCRIPTION
**Pull request by yury-lysogorskiy**

### CODE REVIEW

**Comment by jan-janssen** [comment]
I am going to merge this once `place` is available

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:23]: Please use the [`ImportAlarm`](https://github.com/pyiron/pyiron_base/blob/3295218d87dc1518d6073d9916a859b4e424c4be/pyiron_base/generic/util.py#L238). In your case I guess the decorator would then belong down with `get_{current/final}_potential` and tightens up a smidgen of code duplication.

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:37]: Please use `InputList` instead. The only immediate change I foresee is that it facilitates `.key` functionality instead of `['key']`, but it also helps future-proof the class against the day when we stop supporting `GenericParameters`.

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:89]: It's not relevant yet, but I guess in the future using `InputList` might also save some effort here as there's an [open PR](https://github.com/pyiron/pyiron_base/pull/151) to allow this class to directly write yaml files. Would probably need some massaging to get it to work here, so no changes needed, just a comment towards the future.

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:81]: If this is the case, is it not safer to directly test if the instance is a `TrainingContainer`? I admit I'm not super familiar with the context, so I recognize this might be a dumb question.

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:79]: Nit: we're transitioning over to a different string format
```suggestion
                raise ValueError(f"Provided structure_data filename ({self.structure_data}) doesn't exists")
```

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:92]: Nit: single-quotes are inconsistent with the rest of the file

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:107]:
```suggestion
                raise ValueError(f"Provided potential filename ({self.input["potential"]}) doesn't exists")
```

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:178]:
```suggestion
            "Config": [[f"pair_style pace\n", "pair_coeff  * * {pot_file_name} {elem}\n"]],
```

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:190]: Nit: Super call fits on one line fine

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:198]: Nit: Super call fits on one line fine

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:217]: Decorate with `@import_alarm` and can remove if/else clause (sim. `get_final_potential`)

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:31]: Docstring? At least one of the module, class, and/or `__init__` needs some kind of documentation.

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [inline pyiron_gpl/pacemaker/pacemaker.py:27]: I think pyiron might already have a logger set up you can hook into? Not sure though, @jan-janssen could probably quickly answer that.

**Review by liamhuber — CHANGES_REQUESTED** [review]
A handful of nits, but nothing serious. Mostly just some advice to loop in some new tools from elsewhere in the codebase to more easily do a couple of the things you're already doing.

At a high level the only concern I have is that there's no associated tests. I know for `pyiron_contrib` this is not a requirement; I'll leave it to you and @jan-janssen to decide what the standard should be for `pyiron_gpl`. I also recognize from context that this is likely for the potential workshop and you might not have the time between now and then to write new tests, so even if you decide to keep this repo covered "later" is a totally fair response 😉

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py]: Done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py]: Done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py:150]: I'm note sure if we settled up with the `TrainingContainer` name, so I wouldprefer (for the moment) to use duck-typing

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py]: done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py]: done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py]: done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py:281]: done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py:34]: done

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (jan-janssen) [inline pyiron_gpl/pacemaker/pacemaker.py:13]:
```suggestion
from pyiron_base import GenericJob, InputList, ImportAlarm, Settings
```

Please do not bind to internal modules of `pyiron_base`

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py]: This moved up, into import section wit ha comment: # set loggers level for WARNING to avoid extra print-outs  because pyace do its own logging settings

File: pyiron_gpl/pacemaker/pacemaker.py
Comment (liamhuber) [reply pyiron_gpl/pacemaker/pacemaker.py:148]: If this is the case then I would double down on my statement that you should directly test if the instance is a training container. If you `from pyiron_foo import TrainingContainer` and `isinstance(self.structure_data, TrainingContainer)` here, then as long as your pyiron dev environment includes both `pyiron_gpl` and `pyiron_foo` even an extremely simple IDE will catch this occurrence during refactor->rename:TrainingContainer. However, if you later rename the *method* `get_pandas` on that object, then it's hit or miss if the developer will see it in the refactoring change list because it will be tucked down in the string comparison section.

I'm assuming `TrainingContainer` already exists somewhere and you mean you might still rename it. If it doesn't even exist yet I'd just replace the entire section of code with a TODO pending the completion of that class.

Comment (yury-lysogorskiy) [reply pyiron_gpl/pacemaker/pacemaker.py:152]: OK, agreed

**Review by liamhuber — APPROVED** [review]

**Comment by jan-janssen** [comment]
@yury-lysogorskiy I would suggest to merge this now, then we can create a new `pyiron_gpl` version and only use already packaged software for the testing rather than pulling specific git branches.

**Pull request merged**
